### PR TITLE
Add OpenID authentication support using a KeyCloak server

### PR DESCRIPTION
**Pull request by fmartingr**
#### Summary
- Allow users to authenticate with Mattermost by using an OpenID provider
- Adds provisioning for Keycloak server in in development mode with a postgresql database bundled in the same server
- A default realm bundled into the load test with some default credentials.
  - An optional import file can be provided, if none is set, the default one is used.
  - Also a database dump can be provided in the form of an URI and the data will be imported into the SQL database.
- An optional parameter to generate users in the Keycloak instance
- Automatic configuration of the Mattermost instance by using the generated Keycloak server with the provided `KeycloakClientID`, `KeycloakClientSecret` and `KeycloakRealmName` (only needed when a custom import realm file is provided if the default values do not match).
- Documentation

### CODE REVIEW

File: deployment/terraform/create.go
Comment (agnivade) [inline deployment/terraform/create.go:232]: All of this is actually done by setting the model.Config struct. See `(t *Terraform) updateAppConfig`. This code should exist inside that method.

File: deployment/terraform/keycloak.go
Comment (agarciamontoro) [inline deployment/terraform/keycloak.go:35]: Can we move this declaration closer to where it's used (below the Install realm file if-else block)?

File: deployment/terraform/keycloak.go
Comment (agarciamontoro) [inline deployment/terraform/keycloak.go:175]: Not sure if this is a dumb question, but: could it happen that the users file exists but that it doesn't contain the users we want? For example, we create a deployment with 10 users and then recreate it with 50.

File: deployment/terraform/keycloak.go
Comment (agarciamontoro) [inline deployment/terraform/keycloak.go:188]: Nit to make the code below a bit more clear
```suggestion
		username := fmt.Sprintf("keycloak-auto-%06d", i)
		password := username // we just use the same string for both
```

File: deployment/terraform/keycloak.go
Comment (agarciamontoro) [inline deployment/terraform/keycloak.go:191]: Oh wow there is no way to create more than one user at a time? :sob:

File: deployment/config.go
Comment (agarciamontoro) [inline deployment/config.go:191]: Should we use an `Enable bool` instead of `InstanceCount int` if we're only allowing one instance?

File: loadtest/user/userentity/actions.go
Comment (agarciamontoro) [reply loadtest/user/userentity/actions.go:46]: I wrote this before reading authOpenID, when I understood that no, that's not possible. So we'll have to live with this one, no problem.

**Comment by agnivade** [comment]
Is this ready to review now?

**Comment by fmartingr** [comment]
> Is this ready to review now?

I have to address Alejandro's comments and make some changes from the work in the LT environment. I will assign you again when that's done.

File: deployment/terraform/assets/cluster.tf
Comment (fmartingr) [reply deployment/terraform/assets/cluster.tf:698]: I agree 😅

File: deployment/terraform/assets/variables.tf
Comment (fmartingr) [reply deployment/terraform/assets/variables.tf:124]: Yeah, this is a leftover. Defaults are set in `config.go` to `0`.

File: deployment/config.go
Comment (fmartingr) [reply deployment/config.go:214]: Yes, this is the data for the realm embedded with the load test if no custom realm is provided.

File: api/agent.go
Comment (fmartingr) [reply api/agent.go:376]: I was getting crazy because I got an error and I didn't knew where it came from... 😓

File: loadtest/user/userentity/actions.go
Comment (fmartingr) [reply loadtest/user/userentity/actions.go:111]: Oh _shirt_. I keep forgetting basic things...

**Comment by fmartingr** [comment]
> * Do we need to expose the settings for Keycloak's DB? I'd lean towards being more opinionated whenever possible so that we expose less settings. Given that we don't want to monitor/load-test this server, can we use a one-size-fits-all instance type? Maybe not only for the DB but also for the Keycloak server? Or do the specs differ a lot when using more users? In any case, I think we can at least hardcode the usage of mysql or postgres here.

> * Why do the docs say that production mode is not implemented? Is there anything missing? Also: do we really need the distinction? I guess development mode is just cheaper to deploy because of the lack of DB, right?

Keycloak provides different commands for development/production servers. `start-dev` starts the server into development mode using a local [h2](https://www.h2database.com/html/main.html) database and it's less picky with the server configuration, while production mode requires https and a proper database. If h2 is anything like sqlite my guess is that we should be able to use it to log in all the users in dev-mode so all code for production mode can be removed, but I'd rather do a test run first in case the database I/O is a bottleneck.

> * Could you add some docs on the logic around the prefix for the usernames? The overwriting of the users file path was a bit confusing for me at first.

Yeah, I'm unsure about that part. I wanted to do it more automagic for the user but it breaks basic flows and I don't like it. I will remove the user generation part because is very slow. It's better and faster to manually generate an import file.

File: deployment/terraform/assets/cluster.tf
Comment (fmartingr) [reply deployment/terraform/assets/cluster.tf:804]: I'm going to fix it to psql for now, until we decide if remove it or not.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go:194]: We could call the API I guess, but there's no guarantee it would be faster.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go:178]: Yeah, I see that hapenning if that's an usual way of using the load test. Alternatively you could remove the file and the keycloak database to start from scratch.

File: loadtest/user/userentity/actions.go
Comment (agarciamontoro) [inline loadtest/user/userentity/actions.go:84]: Should this be login/signup?
```suggestion
	// make a request to the openid login/signup page of mattermost
	resp, err := client.Get(ue.client.URL + "/oauth/openid/" + string(action))
```

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Again, incredible work, very thorough and well documented, thanks! I left some more comments below (99% of them nits, sorry)

I also wanted to ask you for a summary of what assets we'll need once this gets in. If I'm not mistaken:
- DB dump of Keycloak (where is this? are we planning to upload it to S3 as the MM DB dump?)
- Realm file (is this really needed? or is it ok using the default one?)
- Anything else?

Thanks for all your work!

File: api/agent.go
Comment (agnivade) [reply api/agent.go:376]: Guilty as charged :)

File: api/agent.go
Comment (agnivade) [inline api/agent.go:404]: Can be combined to `if parts := strings.Split(username, ":"); len(parts) > 1`

File: api/agent.go
Comment (agnivade) [inline api/agent.go:410]: Why is this needed? I thought just the username would become like "openid:user1"? The email should be untouched right? Also, we need to do length validation of the slice here to prevent panics.

File: deployment/config.go
Comment (agnivade) [inline deployment/config.go:192]: Should it not be "false" by default? The default config is always aimed towards production usage.

File: deployment/config.go
Comment (agnivade) [inline deployment/config.go:214]: `It is used .. `. Common translation mistake :)

File: deployment/terraform/agent.go
Comment (agnivade) [reply deployment/terraform/agent.go:151]: I think we've had a discussion here: https://community.mattermost.com/core/pl/1944im15z3f1pn9thb647kdi5y. I am 0/5 as long as it doesn't break permalinks.

File: deployment/terraform/create.go
Comment (agnivade) [inline deployment/terraform/create.go:696]: I see this `8080` being duplicated in a lot of places. Not for now, but later we'd want to move this to a config and use that value instead.

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:38]: Why this `2> /dev/null`? Why not simply pipe to `wc -l`? Also, please check the error here.

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:34]: No need to add error context here? :P

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:44]: Please move this to a .sql file under assets, copy it to the keycloak node via provisioner like:

```
provisioner "file" {
    source      = var.mattermost_license_file
    destination = "/home/ubuntu/mattermost.mattermost-license"
  }
```

And use `psql -f` to execute from the file.

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:61]: The destination realm file is copied in two places. Let's consolidate into a single variable.

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:96]: Question: would we need to modify this to support 100K users? If so, would it make sense to move it to config? (Out of scope for this PR obviously).

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:111]: Is the `string()` really needed? I don't see it in other places where templates are used.

File: deployment/terraform/keycloak.go
Comment (agnivade) [inline deployment/terraform/keycloak.go:134]: Don't you need to do a `sudo systemctl daemon-reload` before enabling? Usually we do that for other services in the test.

File: deployment/terraform/keycloak.go
Comment (agnivade) [reply deployment/terraform/keycloak.go:194]: +1 to this. We should do a quick check to ensure that # of users in keycloak file is equal to # users in users.txt file. And if possible, have a way to add new users, or fail clearly in the logs. But importantly, there should never be a case of some users partially logging in via keycloak and some via email/pass.

File: deployment/terraform/metrics.go
Comment (agnivade) [reply deployment/terraform/metrics.go:165]: This seems wrong. 8080 is the port of the keycloak server. Is it also the same port where it exports metrics? That'd be weird.

File: deployment/terraform/strings.go
Comment (agnivade) [inline deployment/terraform/strings.go:51]: typo

File: loadtest/user/userentity/user.go
Comment (agnivade) [inline loadtest/user/userentity/user.go:37]: Should this be `AuthenticationTypeEmailPass`? 0/5

**Review by agnivade — COMMENTED** [review]
Gave it a first pass. Looks to be in good shape. Will give it a deeper dive after that.

File: api/agent.go
Comment (fmartingr) [reply api/agent.go:410]: Since the `users.txt` file works with `email password` lines, what this does is parse if theres a `:` in the email to check for the auth service, hence parsing: `authservice:email password`. But in order to properly set the email I need to discard the `authservice:` part.

File: deployment/config.go
Comment (fmartingr) [reply deployment/config.go:192]: I left it as `false` because the difference development vs production means. If I use the production option I have to disable some of the features it enables manually (like enforced hostnames, SSL, local cache). I could go that route but in the end we would have the same result but with more work. That is, unless we want to really run it in production mode with all security features enforced as they should be.

File: deployment/terraform/agent.go
Comment (fmartingr) [reply deployment/terraform/agent.go:151]: Is there an straightforward way for me to check that permalinks are working?

Comment (agarciamontoro) [reply deployment/terraform/agent.go:151]: Changing the agents host will not break permalinks: the problem was with the app nodes trying to retrieve the data from the permalink. As long as we don't touch that, permalinks will keep working.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go:38]: I was using that to avoid errors here, but piping to `wc -l` is way more elegant. :thumbsup:

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go:134]: You are right, though I have created environments from scratch that started the service correctly, but a `daemon-reload` should be required. Mhmmm. Maybe is only required for changes but not for new units? I need to check this.

File: deployment/terraform/metrics.go
Comment (fmartingr) [reply deployment/terraform/metrics.go:165]: I forgot the path, but yeah, [they expose metrics in the same port](https://www.keycloak.org/server/configuration-metrics).

File: api/agent.go
Comment (agnivade) [reply api/agent.go]: I thought the authservice is just a prefix for the username right? Or is it for both?

File: deployment/terraform/keycloak.go
Comment (agnivade) [reply deployment/terraform/keycloak.go:97]: Ok yeah that's a good point. I guess we will see in our big test.

**Review by agnivade — APPROVED** [review]
Fantastic work! 50 commits! :scream:

**Comment by agnivade** [comment]
Blocking merge because I want to test this out manually by myself.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go]: Changed it to create the number of users setup in the configuration. If the number is higher it will create more users accordingly, using the number of lines in the `users.txt` file as the source. It will skip user creation if the number of lines is the same or higher than the expected number of users.

File: api/agent.go
Comment (fmartingr) [reply api/agent.go]: It's used in the users txt file only, but ignored for login.

Unsure how the `:` would affect login in, and I selected that symbol so we easily know in code if there's a custom authservice setup. This could be changed to use `openid-username-00001` and check if the first part split by `-` matches an authservice but it seemed more complicated in my head. Would love your fresh eyes input on this.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go]: Mhmmm, how do I keep this in assets and then assign it to a variable if it's bundled in the binary?

File: deployment/config.go
Comment (agnivade) [inline deployment/config.go:201]: We are moving everything to the c7 series. Let's do it here as well.

File: deployment/terraform/agent.go
Comment (agnivade) [inline deployment/terraform/agent.go:27]:
```go
	if t.config.SiteURL != "" {
		if t.output.HasProxy() {
			url = t.config.SiteURL
		} else {
			url = t.config.SiteURL + ":8065"
		}
	} else {
		if t.output.HasProxy() {
			url = t.output.Proxy.PrivateIP
		} else {
			url = t.output.Instances[0].PrivateIP + ":8065"
		}
	}
```

**Review by agnivade — APPROVED** [review]
New commits look good to me. The only small thing is that let's change the default keycloak instance spec to `c7i` series. Thanks again for all the hard work!

File: api/agent.go
Comment (streamer45) [inline api/agent.go:411]: I think this could become a util function.

File: loadtest/user/userentity/user.go
Comment (streamer45) [reply loadtest/user/userentity/user.go:37]: +1

File: loadtest/user/userentity/actions.go
Comment (streamer45) [inline loadtest/user/userentity/actions.go:121]: Do we need to do this here? I'd expect this method to purely authenticate (e.g. set the token) and then deal with everything else in `SignUp/Login`.

File: deployment/terraform/keycloak.go
Comment (streamer45) [inline deployment/terraform/keycloak.go:100]: Maybe we could use a template file as we do with other plaintext assets.

File: deployment/terraform/keycloak.go
Comment (streamer45) [inline deployment/terraform/keycloak.go:235]: Why would `Stat` return an error if the file exists?

**Review by streamer45 — COMMENTED** [review]
Very quickly skimmed through this. Looks great overall. Left some comments.

**Review by agarciamontoro — APPROVED** [review]
Agian, amazing work, @fmartingr. Nothing else to add. Thank you!

File: api/agent.go
Comment (fmartingr) [reply api/agent.go:411]: I'm going to make a note of this but refrain from modifying too much logic until this is merged now that we have this working. I can address the rest of the comments in individual PRs later.

File: deployment/terraform/dump.go
Comment (fmartingr) [reply deployment/terraform/dump.go:174]: Good catch.

File: deployment/terraform/keycloak.go
Comment (fmartingr) [reply deployment/terraform/keycloak.go:235]: I think this was supposed to be `os.IsNotExist` in a previous apporach or something else with a different logic and I just inherited it here, this doesn't make any sense at the moment.

**Comment by agnivade** [comment]
Merge it at your convenience. :smile:

File: loadtest/user/userentity/actions.go
Comment (fmartingr) [reply loadtest/user/userentity/actions.go:121]: There's a mix of logic here, is either I get the user and set it here, or I get the user by username later and set it in the other method. I've set a TODO to change the logic in a separate PR since I don't want to compromise or delay the test we're going to run today.

**Comment by fmartingr** [comment]
> Merge it at your convenience. 😄

@streamer45 I've put comments on heavy logic changes that you mentioned so I can address those on separate PRs so I can merge this for now since it's getting too big and not easy to review. Thank you for reviewing!

**Pull request merged**
